File: workshop-ui/static/js/sections/citations.js
/**
 * Citations section
 */

import { state } from '../state.js';
import { getConfig } from '../utils/api.js';
import { 
    escapeHtml, 
    showError, 
    createMessageElement, 
    addStreamingIndicator, 
    scrollToBottom, 
    updateDebugPanel 
} from '../utils/dom.js';
import { formatFileSize, renderMarkdown } from '../utils/formatting.js';

export function initCitations() {
    initCitationsDropZone();

    document.getElementById('citationsTextInput').addEventListener('input', (e) => {
        const hasText = e.target.value.trim().length > 0;
        const hasFile = state.citationsDocumentBase64 || state.citationsDocumentText;
        document.getElementById('sendCitationsChat').disabled = !hasText && !hasFile;
    });

    document.getElementById('clearCitationsChat').addEventListener('click', () => {
        state.citationsMessages = [];
        document.getElementById('citationsChatMessages').innerHTML = `
            <div class="text-muted text-center py-5">
                <i class="bi bi-quote fs-1"></i>
                <p>Upload a document and ask questions</p>
                <small>Claude will cite specific passages from the document</small>
            </div>
        `;
    });

    document.getElementById('sendCitationsChat').addEventListener('click', sendCitationsChat);

    document.getElementById('citationsChatInput').addEventListener('keydown', (e) => {
        if (e.key === 'Enter' && !e.shiftKey) {
            e.preventDefault();
            sendCitationsChat();
        }
    });
}

function initCitationsDropZone() {
    const dropZone = document.getElementById('citationsDropZone');
    const fileInput = document.getElementById('citationsFileInput');

    dropZone.addEventListener('click', () => fileInput.click());

    dropZone.addEventListener('dragover', (e) => {
        e.preventDefault();
        dropZone.classList.add('drag-over');
    });

    dropZone.addEventListener('dragleave', () => {
        dropZone.classList.remove('drag-over');
    });

    dropZone.addEventListener('drop', async (e) => {
        e.preventDefault();
        dropZone.classList.remove('drag-over');
        const file = e.dataTransfer.files[0];
        if (file) {
            await processCitationsFile(file);
        }
    });

    fileInput.addEventListener('change', async (e) => {
        const file = e.target.files[0];
        if (file) {
            await processCitationsFile(file);
        }
        fileInput.value = '';
    });
}

async function processCitationsFile(file) {
    const ext = file.name.split('.').pop().toLowerCase();

    let mediaType;
    let isText = false;

    switch (ext) {
        case 'pdf':
            mediaType = 'application/pdf';
            break;
        case 'txt':
        case 'md':
            mediaType = 'text/plain';
            isText = true;
            break;
        case 'html':
            mediaType = 'text/html';
            isText = true;
            break;
        default:
            alert('Unsupported file type. Please use PDF, TXT, MD, or HTML.');
            return;
    }

    try {
        if (isText) {
            const text = await file.text();
            state.citationsDocumentText = text;
            state.citationsDocumentBase64 = null;
        } else {
            const arrayBuffer = await file.arrayBuffer();
            const base64 = btoa(String.fromCharCode(...new Uint8Array(arrayBuffer)));
            state.citationsDocumentBase64 = base64;
            state.citationsDocumentText = null;
        }

        state.citationsDocumentType = mediaType;
        state.citationsDocumentTitle = file.name;
        state.citationsDocument = file;

        renderCitationsUploadedDoc();
        document.getElementById('sendCitationsChat').disabled = false;

    } catch (error) {
        console.error('Error processing file:', error);
        alert('Failed to process file: ' + error.message);
    }
}

function renderCitationsUploadedDoc() {
    const container = document.getElementById('citationsUploadedDoc');

    if (!state.citationsDocument && !state.citationsDocumentText) {
        container.innerHTML = '';
        return;
    }

    const isPdf = state.citationsDocumentType === 'application/pdf';
    const icon = isPdf ? 'bi-file-pdf' : 'bi-file-text';
    const typeClass = isPdf ? 'pdf' : 'text';

    const name = state.citationsDocumentTitle || 'Document';
    const size = state.citationsDocument ? formatFileSize(state.citationsDocument.size) : '';

    container.innerHTML = `
        <div class="uploaded-doc ${typeClass}">
            <i class="bi ${icon}"></i>
            <div class="doc-info">
                <div class="doc-name">${escapeHtml(name)}</div>
                <div class="doc-size">${size}</div>
            </div>
            <span class="remove-doc" onclick="removeCitationsDoc()">
                <i class="bi bi-x-lg"></i>
            </span>
        </div>
    `;
}

function removeCitationsDoc() {
    state.citationsDocument = null;
    state.citationsDocumentBase64 = null;
    state.citationsDocumentText = null;
    state.citationsDocumentTitle = 'Document';
    renderCitationsUploadedDoc();
    document.getElementById('sendCitationsChat').disabled = true;
}

async function sendCitationsChat() {
    if (state.isStreaming) return;

    const input = document.getElementById('citationsChatInput');
    const message = input.value.trim();
    if (!message) return;

    const textInput = document.getElementById('citationsTextInput');
    const textTitle = document.getElementById('citationsTextTitle');

    if (!state.citationsDocumentBase64 && !state.citationsDocumentText) {
        if (textInput.value.trim()) {
            state.citationsDocumentText = textInput.value.trim();
            state.citationsDocumentTitle = textTitle.value.trim() || 'Pasted Text';
            state.citationsDocumentType = 'text/plain';
            document.getElementById('sendCitationsChat').disabled = false;
        } else {
            alert('Please upload a document or paste text first');
            return;
        }
    }

    input.value = '';

    state.citationsMessages.push({ role: 'user', content: message });

    const container = document.getElementById('citationsChatMessages');
    if (container.querySelector('.text-muted.text-center')) {
        container.innerHTML = '';
    }

    await streamCitationsChat('/api/citations/chat', {
        config: getConfig(),
        messages: state.citationsMessages,
        document_base64: state.citationsDocumentBase64,
        document_text: state.citationsDocumentText,
        document_type: state.citationsDocumentType,
        document_title: state.citationsDocumentTitle
    }, {
        containerId: 'citationsChatMessages',
        debugPanelId: 'citationsDebug',
        messagesKey: 'citationsMessages'
    });
}

async function streamCitationsChat(endpoint, body, options) {
    const { containerId, debugPanelId } = options;
    const container = document.getElementById(containerId);

    state.isStreaming = true;

    if (body.messages && body.messages.length > 0) {
        const lastUserMsg = body.messages[body.messages.length - 1];
        if (lastUserMsg.role === 'user') {
            container.appendChild(createMessageElement('user', lastUserMsg.content));
        }
    }

    const streamingDiv = addStreamingIndicator(container);
    let responseText = '';
    let citations = [];

    try {
        const response = await fetch(endpoint, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(body)
        });

        const reader = response.body.getReader();
        const decoder = new TextDecoder();

        while (true) {
            const { done, value } = await reader.read();
            if (done) break;

            const chunk = decoder.decode(value);
            const lines = chunk.split('\n');

            for (const line of lines) {
                if (line.startsWith('data: ')) {
                    try {
                        const data = JSON.parse(line.slice(6));

                        switch (data.type) {
                            case 'debug_request':
                                updateDebugPanel(debugPanelId, { request: data.data }, false);
                                break;

                            case 'debug_response':
                                updateDebugPanel(debugPanelId, { response: data.data }, true);
                                break;

                            case 'text_with_citations':
                                responseText = data.data.text;
                                citations = data.data.citations || [];

                                const renderedContent = renderTextWithCitations(responseText, citations);
                                streamingDiv.querySelector('.message-content').innerHTML = renderedContent;
                                scrollToBottom(container);
                                break;

                            case 'error':
                                showError(data.data, containerId);
                                break;

                            case 'done':
                                break;
                        }
                    } catch (e) {
                        console.error('Failed to parse SSE data:', e);
                    }
                }
            }
        }

        streamingDiv.remove();
        if (responseText) {
            state.citationsMessages.push({ role: 'assistant', content: responseText });

            const msgDiv = document.createElement('div');
            msgDiv.className = 'message message-assistant message-with-citations';
            msgDiv.innerHTML = `
                <div class="message-role">Assistant</div>
                <div class="message-content">${renderTextWithCitations(responseText, citations)}</div>
            `;
            container.appendChild(msgDiv);

            msgDiv.setAttribute('data-citations', JSON.stringify(citations));
        }

    } catch (error) {
        streamingDiv.remove();
        showError(error.message, containerId);
    } finally {
        state.isStreaming = false;
    }
}

function renderTextWithCitations(text, citations) {
    if (!citations || citations.length === 0) {
        return renderMarkdown(text);
    }

    let html = renderMarkdown(text);

    const citationRefs = citations.map((c, idx) => {
        const preview = c.cited_text ? c.cited_text.substring(0, 100) + (c.cited_text.length > 100 ? '...' : '') : '';
        return `
            <span class="citation-marker"
                  onclick="showCitationDetail(${idx})"
                  data-citation-index="${idx}"
                  title="${escapeHtml(preview)}">
                [${idx + 1}]
            </span>
        `;
    }).join(' ');

    window.currentCitations = citations;

    return html + `<div class="mt-2"><small class="text-muted">Citations: </small>${citationRefs}</div>`;
}

function showCitationDetail(idx) {
    const citations = window.currentCitations || [];
    const citation = citations[idx];
    if (!citation) return;

    document.getElementById('citationModalDocTitle').textContent = citation.document_title || 'Unknown';
    document.getElementById('citationModalText').textContent = citation.cited_text || 'No text available';

    const pageContainer = document.getElementById('citationModalPageContainer');
    const charContainer = document.getElementById('citationModalCharContainer');

    if (citation.page_number !== undefined) {
        document.getElementById('citationModalPage').textContent = citation.page_number;
        pageContainer.style.display = 'block';
    } else {
        pageContainer.style.display = 'none';
    }

    if (citation.start_char_index !== undefined && citation.end_char_index !== undefined) {
        document.getElementById('citationModalCharRange').textContent = `${citation.start_char_index} - ${citation.end_char_index}`;
        charContainer.style.display = 'block';
    } else {
        charContainer.style.display = 'none';
    }

    const modal = new bootstrap.Modal(document.getElementById('citationDetailModal'));
    modal.show();
}

// Make functions available globally
window.removeCitationsDoc = removeCitationsDoc;
window.showCitationDetail = showCitationDetail;

File: workshop-ui/app/routes/citations.py
"""Citations endpoint."""

import json
from fastapi import APIRouter
from fastapi.responses import StreamingResponse
import anthropic

from ..models.features import CitationsChatRequest
from ..utils.client import get_client
from ..utils.helpers import format_request_for_debug

router = APIRouter()


@router.post("/citations/chat")
async def citations_chat(request: CitationsChatRequest):
    """Chat with Claude using citations on documents."""
    client = get_client(request.config)

    messages = []
    user_content = []

    # Add document with citations enabled
    if request.document_base64:
        user_content.append({
            "type": "document",
            "source": {
                "type": "base64",
                "media_type": request.document_type,
                "data": request.document_base64
            },
            "title": request.document_title,
            "citations": {"enabled": True}
        })
    elif request.document_text:
        user_content.append({
            "type": "document",
            "source": {
                "type": "text",
                "media_type": "text/plain",
                "data": request.document_text
            },
            "title": request.document_title,
            "citations": {"enabled": True}
        })

    # Add previous messages
    for msg in request.messages:
        msg_dict = msg.model_dump()
        if msg_dict["role"] == "user":
            if isinstance(msg_dict["content"], str):
                if not messages:  # First message - include document
                    user_content.append({"type": "text", "text": msg_dict["content"]})
                    messages.append({"role": "user", "content": user_content})
                else:
                    messages.append({"role": "user", "content": msg_dict["content"]})
            else:
                if not messages:
                    user_content.extend(msg_dict["content"] if isinstance(msg_dict["content"], list) else [msg_dict["content"]])
                    messages.append({"role": "user", "content": user_content})
                else:
                    messages.append(msg_dict)
        else:
            messages.append(msg_dict)

    # If no messages were added yet, add just the document
    if not messages and user_content:
        messages.append({"role": "user", "content": user_content})

    params = {
        "model": request.config.model,
        "max_tokens": request.max_tokens,
        "messages": messages
    }
    if request.system:
        params["system"] = request.system

    debug_request = format_request_for_debug("/v1/messages (citations)", params)

    async def generate():
        yield f"data: {json.dumps({'type': 'debug_request', 'data': debug_request})}\n\n"

        try:
            response = client.messages.create(**params)

            response_data = {
                "id": response.id,
                "model": response.model,
                "content": [],
                "stop_reason": response.stop_reason,
                "usage": {
                    "input_tokens": response.usage.input_tokens,
                    "output_tokens": response.usage.output_tokens
                }
            }

            for block in response.content:
                if block.type == "text":
                    text_block = {
                        "type": "text",
                        "text": block.text,
                        "citations": []
                    }

                    if hasattr(block, 'citations') and block.citations:
                        for citation in block.citations:
                            citation_data = {
                                "type": citation.type if hasattr(citation, 'type') else "unknown",
                                "cited_text": citation.cited_text if hasattr(citation, 'cited_text') else "",
                                "document_title": citation.document_title if hasattr(citation, 'document_title') else "",
                                "document_index": citation.document_index if hasattr(citation, 'document_index') else 0,
                            }
                            if hasattr(citation, 'start_char_index'):
                                citation_data["start_char_index"] = citation.start_char_index
                            if hasattr(citation, 'end_char_index'):
                                citation_data["end_char_index"] = citation.end_char_index
                            if hasattr(citation, 'page_number'):
                                citation_data["page_number"] = citation.page_number

                            text_block["citations"].append(citation_data)

                    response_data["content"].append(text_block)
                    yield f"data: {json.dumps({'type': 'text_with_citations', 'data': text_block})}\n\n"

            yield f"data: {json.dumps({'type': 'debug_response', 'data': response_data})}\n\n"
            yield f"data: {json.dumps({'type': 'done'})}\n\n"

        except anthropic.APIError as e:
            error_msg = str(e)
            if "citation" in error_msg.lower() or "document" in error_msg.lower():
                error_msg = f"Citations API Error: {error_msg}\n\nNote: Make sure your document format is supported and citations are properly configured."
            yield f"data: {json.dumps({'type': 'error', 'data': error_msg})}\n\n"
        except Exception as e:
            yield f"data: {json.dumps({'type': 'error', 'data': f'Error: {type(e).__name__}: {str(e)}'})}\n\n"

    return StreamingResponse(generate(), media_type="text/event-stream")
